controllers/vmware/vspherecluster_reconciler_test.go

```diff
@@ -128,34 +136,149 @@ var _ = Describe("Cluster Controller Tests", func() {
+			objects:     []client.Object{},
+			want:        nil,
+			wantErr:     false,
+			featureGate: true,
+		},
+		{
+			name:        "Cluster-Wide: should not find any FailureDomains if only namespaced exist",
+			objects:     []client.Object{zone(namespace.Name, "ns-one", false)},
+			want:        nil,
+			wantErr:     false,
+			featureGate: false,
+		},
+		{
+			name:        "Namespaced: should not find any FailureDomains if only cluster-wide exist",
+			objects:     []client.Object{availabiltyZone("c-one")},
+			want:        nil,
+			wantErr:     false,
+			featureGate: true,
+		},
+		{
+			name:        "Cluster-Wide: should find FailureDomains if only cluster-wide exist",
+			objects:     []client.Object{availabiltyZone("c-one")},
+			want:        failureDomains("c-one"),
+			wantErr:     false,
+			featureGate: false,
+		},
+		{
+			name:        "Namespaced: should find FailureDomains if only namespaced exist",
+			objects:     []client.Object{zone(namespace.Name, "ns-one", false)},
+			want:        failureDomains("ns-one"),
+			wantErr:     false,
+			featureGate: true,
+		},
+		{
+			name:        "Cluster-Wide: should only find cluster-wide FailureDomains if both types exist",
+			objects:     []client.Object{availabiltyZone("c-one"), zone(namespace.Name, "ns-one", false)},
+			want:        failureDomains("c-one"),
+			wantErr:     false,
+			featureGate: false,
+		},
+		{
+			name:        "Namespaced: should only find namespaced FailureDomains if both types exist",
+			objects:     []client.Object{availabiltyZone("c-one"), zone(namespace.Name, "ns-one", false)},
+			want:        failureDomains("ns-one"),
+			wantErr:     false,
+			featureGate: true,
+		},
+		{
+			name: "Namespaced: should only find non-deleting namespaced FailureDomains",
+			objects: []client.Object{
+				availabiltyZone("c-one"),
+				zone(namespace.Name, "ns-one", false),
+				zone(namespace.Name, "ns-two", false),
+				zone(namespace.Name, "ns-three", false),
+				zone(namespace.Name, "ns-four", true),
+			},
+			want:        failureDomains("ns-one", "ns-two", "ns-three"),
+			wantErr:     false,
+			featureGate: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &ClusterReconciler{
+				Client: fake.NewClientBuilder().
+					WithScheme(scheme).
+					WithObjects(append([]client.Object{namespace}, tt.objects...)...).
+					Build(),
+			}
+			defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.NamespaceScopedZone, tt.featureGate)()
+			got, err := r.getFailureDomains(ctx, namespace.Name)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ClusterReconciler.getFailureDomains() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ClusterReconciler.getFailureDomains() = %v, want %v", got, tt.want)
+			}
 		})
-	})
-})
+	}
+}
+
+func availabiltyZone(name string) *topologyv1.AvailabilityZone {
```

Comment:
```suggestion
func availabilityZone(name string) *topologyv1.AvailabilityZone {
```

Reply:
Good eye:)

Reply:
Intellij spell checker :)

Reply:
my bad

Reply:
@zhanggbj Not important, but I think this one was not fixed

Reply:
Oh yes, will fix it in next patch.